### PR TITLE
[SPIRV] Add support hlsl export function attribute

### PR DESCRIPTION
**Pull request by jiaolu**
DXIL could export functions with export attribute while spirv would
ignore such functions. Currently glsl spirv does not have linkage
export decorate. so the exported function would be treated as normal
function while add dummy entry point according to the spirv validations

### CODE REVIEW

**Comment by jiaolu** [comment]
Hi, @jaebaek @ehsannas, do you any comments on this pull request?
Thanks

**Comment by ehsannas** [comment]
I'd been out of office for a few days. I'll take a look by the end of the week. Thanks

**Comment by ehsannas** [comment]
@jiaolu  Can you please explain how you want to use this exported function in SPIR-V? Are you using `spirv-link`? do you have a working example that we can look at?

**Comment by jiaolu** [comment]
> @jiaolu Can you please explain how you want to use this exported function in SPIR-V? Are you using `spirv-link`? do you have a working example that we can look at?

there is internal project we need to use the hlsl converted spv as lib input .
we don't do spirv-link per se, but the the spv would be converted to the Llvm IR and to be linked to other llvm IR generated from app.

**Comment by jiaolu** [comment]
> @jiaolu Can you please explain how you want to use this exported function in SPIR-V? Are you using `spirv-link`? do you have a working example that we can look at?

currently our internal project use glslangValidator to generate spv with --keep-uncalled-function option. I am intended to switch to dxc with better hlsl support , language feature, spv validation, etc

**Review by ehsannas — COMMENTED** [review]
The idea of the wrapper function is that the main entry point of a vulkan shader should be of the 
```cpp
void entryPointFunction (void) {...}
```
signature. Meaning, it should not take any arguments and return nothing. That's why for regular shaders, we wrap their entry point function in a `void fn(void)` function.

In this case
a) why are you adding a wrapper function? Is it necessary?
b) if it is necessary, your wrapper function's signature is the same as the inner main function signature, so it doesn't serve any purpose. For example, for the `fn.export.hlsl` file, I get:

```cpp
       %main = OpFunction %int None %12
      %color = OpFunctionParameter %_ptr_Function_v4float
         %17 = OpLabel
         %18 = OpFunctionCall %int %src_main %color
               OpReturnValue %18
               OpFunctionEnd
   %src_main = OpFunction %int None %12
```

Notice that the signature of `main` and `src_main` are the same.

**Comment by jiaolu** [comment]
Updated to remove emitEntryFunctionWrapperForLibrary.

**Comment by ehsannas** [comment]
@jiaolu I've made updates to your PR.

Explanation about my changes:
* The reason you were adding a "dummy" entrypoint was that spirv-val would error out that a shader without an entrypoint is invalid.
* The proper fix for this is to use the `Linkage` capability.
  * According to the SPIR-V spec, a shader with the `Linkage` capability wihtout an entrypoint is valid.
* Note: `Linkage` capability is not allowed to be used by the Vulkan target environment
  * So, we need to use `-fspv-target-env=universal1.5` to avoid getting a spirv-val error about using `Linkage`
  * The final SPIR-V coming out of your tool (after merging all SPIR-V) should remove the `Linkage` capability so that the result is a valid shader for Vulkan.

 Please let me know if it satisfies your needs.

**Comment by ehsannas** [comment]
Your test shader still works. See: https://github.com/microsoft/DirectXShaderCompiler/pull/3234/files#diff-34a81f6f0dc029f81820b602e839e6fc5802084f61550d6ee47fd6afe8bfdad4R1

**Comment by jiaolu** [comment]
thanks, removing dummyEntry is all the better if possible

**Comment by jiaolu** [comment]
what's spv looks like?

**Comment by ehsannas** [comment]
For your shader:
```cpp
// Run: %dxc -T lib_6_3 -fspv-target-env=universal1.5
RWBuffer< float4 > output : register(u1);
export int MyExportedFunction(inout float4 color) {
  output[0] = color;
  return 1;
}
```

You will get something like this:
```
; SPIR-V
; Version: 1.5
; Generator: Google spiregg; 0
; Bound: 18
; Schema: 0
               OpCapability SampledBuffer
               OpCapability ImageBuffer
               OpCapability Shader
               OpCapability Linkage
               OpMemoryModel Logical GLSL450
               OpSource HLSL 630
               OpName %type_buffer_image "type.buffer.image"
               OpName %output "output"
               OpName %MyExportedFunction "MyExportedFunction"
               OpName %color "color"
               OpName %bb_entry "bb.entry"
               OpDecorate %output DescriptorSet 0
               OpDecorate %output Binding 1
       %uint = OpTypeInt 32 0
     %uint_0 = OpConstant %uint 0
        %int = OpTypeInt 32 1
      %int_1 = OpConstant %int 1
      %float = OpTypeFloat 32
%type_buffer_image = OpTypeImage %float Buffer 2 0 0 2 Rgba32f
%_ptr_UniformConstant_type_buffer_image = OpTypePointer UniformConstant %type_buffer_image
    %v4float = OpTypeVector %float 4
%_ptr_Function_v4float = OpTypePointer Function %v4float
          %9 = OpTypeFunction %int %_ptr_Function_v4float
     %output = OpVariable %_ptr_UniformConstant_type_buffer_image UniformConstant
       %MyExportedFunction = OpFunction %int None %9
      %color = OpFunctionParameter %_ptr_Function_v4float
   %bb_entry = OpLabel
         %15 = OpLoad %v4float %color
         %16 = OpLoad %type_buffer_image %output
               OpImageWrite %16 %uint_0 %15 None
               OpReturnValue %int_1
               OpFunctionEnd
```

**Comment by jiaolu** [comment]
thanks, it looks good to me.

**Comment by jiaolu** [comment]
@ehsannas,  code is rebased,  . Is there anything left to do?

**Comment by ehsannas** [comment]
@jiaolu your force push has removed all of my code. Please refrain from force-push in the future.

**Comment by ehsannas** [comment]
Luckily I still have the code available on my local machine. I can't approve/merge the code since I have contributed to this PR. We need to wait for a code review and approval.

File: tools/clang/lib/SPIRV/SpirvEmitter.cpp
Comment (jaebaek) [inline tools/clang/lib/SPIRV/SpirvEmitter.cpp:10474]: Adding a comment to explain why `Library` must be `GLCompute` would be helpful.

**Review by jaebaek — APPROVED** [review]
Mostly looks good to me with three minor issues.

File: tools/clang/lib/SPIRV/SpirvEmitter.cpp
Comment (ehsannas) [reply tools/clang/lib/SPIRV/SpirvEmitter.cpp]: Good catch. This was leftover from original PR. I've removed it now.

**Comment by jiaolu** [comment]
> @jiaolu your force push has removed all of my code. Please refrain from force-push in the future.

sorry about that.  I did not expect this

**Comment by ehsannas** [comment]
@jiaolu No worries! Good to keep in mind for future PRs. Having multiple commits makes it easier to review and see changes from the last time the code was updated too. Thanks

**Pull request merged**
